FabricObserver/Observers/AppObserver.cs

```diff
@@ -2455,11 +2689,120 @@ private async Task SetDeployedApplicationReplicaOrInstanceListAsync(Uri applicat
+            }
+
+            if (string.IsNullOrWhiteSpace(servicePkgName))
+            {
+                ObserverLogger.LogInfo($"Invalid value for {nameof(servicePkgName)}: {servicePkgName}. Exiting TupleGetResourceGovernanceInfo.");
+                return (false, 0);
+            }
+
+            // Don't waste cycles with XML parsing if you can easily get a hint first..
+            if (!appManifestXml.Contains("<ResourceGovernancePolicy "))
+            {
+                return (false, 0);
+            }
+
+            // Safe XML pattern - *Do not use LoadXml*.
+            var appManifestXdoc = new XmlDocument { XmlResolver = null };
+
+            using (var sreader = new StringReader(appManifestXml))
+            {
+                using (var xreader = XmlReader.Create(sreader, new XmlReaderSettings { XmlResolver = null }))
+                {
+                    appManifestXdoc?.Load(xreader);
+
+                    ObserverLogger.LogInfo("Completed TupleGetResourceGovernanceInfo.");
+                    return TupleGetResourceGovernanceInfoFromAppManifest(ref appManifestXdoc, servicePkgName);
+                }
+            }
+        }
+
+        private (bool IsRGEnabled, double MemoryLimitMb) TupleGetResourceGovernanceInfoFromAppManifest(ref XmlDocument xDoc, string servicePkgName)
+        {
+            ObserverLogger.LogInfo("Starting TupleGetResourceGovernanceInfoFromAppManifest.");
+           
+            if (xDoc == null)
+            {
+                ObserverLogger.LogInfo("Completed TupleGetResourceGovernanceInfoFromAppManifest: xDoc == null.");
+                return (false, 0);
+            }
+
+            try
+            {
+                // Ensure you get the correct manifest import for specified service package name. There will generally be multiple RG limits set per application (so, per service settings).
+                var sNode = 
+                    xDoc.DocumentElement?.SelectSingleNode(
+                        $"//*[local-name()='ServiceManifestImport'][*[local-name()='ServiceManifestRef' and @*[local-name()='ServiceManifestName' and . ='{servicePkgName}']]]");
+                
+                if (sNode == null)
+                {
+                    ObserverLogger.LogInfo($"Completing TupleGetResourceGovernanceInfoFromAppManifest: Missing ServiceManifestImport for {servicePkgName}.");
+                    return (false, 0);
+                }
+
+                XmlNodeList childNodes = sNode.ChildNodes;
+                XmlNode rgNode = null;
+
+                foreach (XmlNode node in childNodes)
+                {
+                    if (node.Name == "Policies")
+                    {
+                        foreach (XmlNode polNode in node.ChildNodes)
+                        {
+                            if (polNode.Name == "ResourceGovernancePolicy")
+                            {
+                                rgNode = polNode;
+                                break;
```

Comment:
Important: if we break here, we can skip RG limit for some code package in case service package has multiple code packages.

Can we please have a support for that as well - to have multiple code packages needed for a single replica to run?

Reply:
Yes. Good catch. The foreach should be expanded to cover all Code packages specified.

Reply:
This is actually a relatively hard problem to solve based on the way FO detects host processes via replica queries. In the case where another code package is a helper exe, FO doesn't detect that today, so I will have work on the problem to support this case. It is non-trivial (given the current impl for process lookups) but certainly doable. Will ship a solution as soon as possible..

Reply:
I have an initial solution to this. Will push Monday. On call Tues-Fri, so there will be a delay in final impl..

Reply:
Yeah, we might need to introduce monitoring for all the helper code packages with special logic for logging. How about we ship current solution without it (because of incoming on-call obligations), measure how many helper code packages do we usually have and follow up with support for those later?

Reply:
The current solution treats all code packages as individual services, so any supported metric will be measured. I think it can make this release.

Reply:
This feature is captured in this Work Item/Feature Request: https://github.com/microsoft/service-fabric-observer/issues/248